From patchwork Tue May 18 12:00:31 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 12264681
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FORGED_FROMDOMAIN,FREEMAIL_FROM,
	HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id A4DFDC433ED
	for <kvm@archiver.kernel.org>; Tue, 18 May 2021 12:01:41 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 78BFA60FE9
	for <kvm@archiver.kernel.org>; Tue, 18 May 2021 12:01:41 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1348982AbhERMCz (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 18 May 2021 08:02:55 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:49392 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S233662AbhERMCu (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 18 May 2021 08:02:50 -0400
Received: from mail-pj1-x102a.google.com (mail-pj1-x102a.google.com
 [IPv6:2607:f8b0:4864:20::102a])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 65F60C061573;
        Tue, 18 May 2021 05:01:32 -0700 (PDT)
Received: by mail-pj1-x102a.google.com with SMTP id
 j6-20020a17090adc86b02900cbfe6f2c96so1419317pjv.1;
        Tue, 18 May 2021 05:01:32 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id;
        bh=em3y75NXotaLU/AKtI4LMXPcEnf5Ld1kv/mHHILm70c=;
        b=V/p4lH/6d5UguGqICWMegLeErmFH7naXSe4DlY8l+O9zRNBHKkp8LUKd2lN4M2qCZ/
         Trfkad+GRBI0RamLpWK3ZyCmXmEqFbV6NJePLgTFCNJjgMHwGWgCRtKt00TIYG0joO4k
         anoh6qGD6DuPuv5gV77s2MQLWPIO5Kx3XOTx+XtklL587DU3TGpUjnvGY3qK9Ntrg/IG
         /cYtXr8tfUYW2Qkv/HpRerACyiOJdAlfcp0KaWDk0Yvlo+LjdkWl41LDnazF1MNDs5Bw
         HlfGiR1kq7LWa2iaXyCbebVh9T/ue0MEoJ69JYBlgMZ0z8HSU8xW2H/p7ngeQ7Cqy9ni
         eJLQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id;
        bh=em3y75NXotaLU/AKtI4LMXPcEnf5Ld1kv/mHHILm70c=;
        b=osG/hjTWDXsgCY9PgdzdlhEM9OlZzIHq+F1PmWpYyzmOZKKuoKD9Q13auMHnCGoK5B
         2yM0hc5Cgo+8QjUGtRE59skDh8Mua0FsTb2pnPuq+ZGiemUIqri6cDoFO9ITdwpcl/qh
         Hw32QwOH0dlztBCl3K/h6wpA8cjnEI7L5NyTWAghXn3yioj7jfoF2Y1KVq/VoHjBNUTv
         AioiQY4z/eJvsiZgFrHAYgCirjEJiKleEXpczWCvShONyqWPM4+wyhNugJDqw1InUX+Y
         gKb0JH55oSusiJTUz9cm0eeqwzAh3OrWX3E5QD20/trBoFxKjrOD6j7FchhE6VVUA6P6
         xr0w==
X-Gm-Message-State: AOAM5329CEqUv76cUaLUwUFubKkIzjj9ZkB5r3xycnMQqU2EEYVvUir9
        SgwLLUiehss3a4CVC+EJbru/M3OX+Q0=
X-Google-Smtp-Source: 
 ABdhPJzuaIu1kDieJYdkyutSSE1hKoGm9LjVD+Vi0D2+YlF331ZxLvgY6I3C4fUQhe3dRm4RAN44MQ==
X-Received: by 2002:a17:902:f203:b029:f0:d225:c6e4 with SMTP id
 m3-20020a170902f203b02900f0d225c6e4mr4271634plc.0.1621339291776;
        Tue, 18 May 2021 05:01:31 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.53])
        by smtp.googlemail.com with ESMTPSA id
 l20sm12757394pjq.38.2021.05.18.05.01.27
        (version=TLS1_2 cipher=ECDHE-ECDSA-AES128-GCM-SHA256 bits=128/128);
        Tue, 18 May 2021 05:01:31 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
        Sean Christopherson <seanjc@google.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>,
        Ben Segall <bsegall@google.com>,
        Venkatesh Srinivas <venkateshs@chromium.org>,
        David Matlack <dmatlack@google.com>,
        Paul Mackerras <paulus@ozlabs.org>,
        Suraj Jitindar Singh <sjitindarsingh@gmail.com>
Subject: [PATCH v4 1/5] KVM: exit halt polling on need_resched() for both
 book3s and generic halt-polling
Date: Tue, 18 May 2021 05:00:31 -0700
Message-Id: <1621339235-11131-1-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Wanpeng Li <wanpengli@tencent.com>

Inspired by commit 262de4102c7bb8 (kvm: exit halt polling on need_resched()
as well), CFS_BANDWIDTH throttling will use resched_task() when there is just
one task to get the task to block. It was likely allowing VMs to overrun their
quota when halt polling. Due to PPC implements an arch specific halt polling
logic, we should add the need_resched() checking there as well. This
patch adds a helper function that to be shared between book3s and generic
halt-polling loop.

Reviewed-by: David Matlack <dmatlack@google.com>
Reviewed-by: Venkatesh Srinivas <venkateshs@chromium.org>
Cc: Ben Segall <bsegall@google.com>
Cc: Venkatesh Srinivas <venkateshs@chromium.org>
Cc: Jim Mattson <jmattson@google.com> 
Cc: David Matlack <dmatlack@google.com>
Cc: Paul Mackerras <paulus@ozlabs.org>
Cc: Suraj Jitindar Singh <sjitindarsingh@gmail.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
v3 -> v4:
 * rename to kvm_vcpu_can_poll
v2 -> v3:
 * add a helper function
v1 -> v2:
 * update patch description

 arch/powerpc/kvm/book3s_hv.c | 2 +-
 include/linux/kvm_host.h     | 2 ++
 virt/kvm/kvm_main.c          | 8 ++++++--
 3 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/arch/powerpc/kvm/book3s_hv.c b/arch/powerpc/kvm/book3s_hv.c
index 28a80d240b76..7360350e66ff 100644
--- a/arch/powerpc/kvm/book3s_hv.c
+++ b/arch/powerpc/kvm/book3s_hv.c
@@ -3936,7 +3936,7 @@ static void kvmppc_vcore_blocked(struct kvmppc_vcore *vc)
 				break;
 			}
 			cur = ktime_get();
-		} while (single_task_running() && ktime_before(cur, stop));
+		} while (kvm_vcpu_can_poll(cur, stop));
 
 		spin_lock(&vc->lock);
 		vc->vcore_state = VCORE_INACTIVE;
diff --git a/include/linux/kvm_host.h b/include/linux/kvm_host.h
index 2f34487e21f2..ba682f738a25 100644
--- a/include/linux/kvm_host.h
+++ b/include/linux/kvm_host.h
@@ -1583,4 +1583,6 @@ static inline void kvm_handle_signal_exit(struct kvm_vcpu *vcpu)
 /* Max number of entries allowed for each kvm dirty ring */
 #define  KVM_DIRTY_RING_MAX_ENTRIES  65536
 
+bool kvm_vcpu_can_poll(ktime_t cur, ktime_t stop);
+
 #endif
diff --git a/virt/kvm/kvm_main.c b/virt/kvm/kvm_main.c
index 6b4feb92dc79..62522c12beba 100644
--- a/virt/kvm/kvm_main.c
+++ b/virt/kvm/kvm_main.c
@@ -2945,6 +2945,11 @@ update_halt_poll_stats(struct kvm_vcpu *vcpu, u64 poll_ns, bool waited)
 		vcpu->stat.halt_poll_success_ns += poll_ns;
 }
 
+bool kvm_vcpu_can_poll(ktime_t cur, ktime_t stop)
+{
+	return single_task_running() && !need_resched() && ktime_before(cur, stop);
+}
+
 /*
  * The vCPU has executed a HLT instruction with in-kernel mode enabled.
  */
@@ -2973,8 +2978,7 @@ void kvm_vcpu_block(struct kvm_vcpu *vcpu)
 				goto out;
 			}
 			poll_end = cur = ktime_get();
-		} while (single_task_running() && !need_resched() &&
-			 ktime_before(cur, stop));
+		} while (kvm_vcpu_can_poll(cur, stop));
 	}
 
 	prepare_to_rcuwait(&vcpu->wait);

From patchwork Tue May 18 12:00:32 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 12264689
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FORGED_FROMDOMAIN,FREEMAIL_FROM,
	HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 910B1C43461
	for <kvm@archiver.kernel.org>; Tue, 18 May 2021 12:01:48 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 65D4B60231
	for <kvm@archiver.kernel.org>; Tue, 18 May 2021 12:01:48 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1348993AbhERMDB (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 18 May 2021 08:03:01 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:49404 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1348968AbhERMCy (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 18 May 2021 08:02:54 -0400
Received: from mail-pf1-x42d.google.com (mail-pf1-x42d.google.com
 [IPv6:2607:f8b0:4864:20::42d])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 158B8C061756;
        Tue, 18 May 2021 05:01:36 -0700 (PDT)
Received: by mail-pf1-x42d.google.com with SMTP id w1so4564338pfu.0;
        Tue, 18 May 2021 05:01:36 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=+j3s61DfilBbdVnALbykcBci/7JJ6WTH2JZ9k+XRHnU=;
        b=JoYONhNC/crNh3EiW+c93OtVBpj0HK7UEZNWa7+op6Y46s/45mG2LQxvjZeOb20Zp6
         yz/dezxs45qjMBA4KvEMn62iF1Uonlb37c0JTrddjLa4CnVK+6o5DGiSGNWCiFAoV2A2
         YRaJrnBwb+UlSs+z5kc6uxUNQ+6a1IEE2BWsbVIERlQqZkjGtFcUFJ3At/gBJkdwMD3o
         cXUREVb2h0Z6tHaIatje+Z84bYjuGafvg/noM/WmxV+vPchX3+QasW3HzrVgHVZu1i3Z
         zbnu1XuziJY6RTFDsqP5CQe1x3hgEgP0di1jZRN/1LNaXAc7M61pB8x601joEt5Ws5cG
         LCdw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references;
        bh=+j3s61DfilBbdVnALbykcBci/7JJ6WTH2JZ9k+XRHnU=;
        b=mewptrc63s16TZcx1Ok9bNR3zvp25kcKnemG1M6TEAbta3VmBvj4PCQrOzccQixq1M
         oejhKeKftCAOerRHalCV9HXbbrPPsByGAI16JsFzPcRRGrlYY2AKBtym4JiGJgBrBGao
         mYN0YNScEv71uNk4nCc5eF7gyhlhkurcYGvIzkqU6sOlhkt1s8Be6pPlNhZ/w7FwcDNK
         ob42YwCwdtTZxL7vz8oWSQmGnyze8z3ki3jAIbG/y5/4cj4xBizhNj+d5OIh3MHMviG0
         Z8pSYhNxcpFee6F/7lylOFNrimuEvAIMzCBQuHWIsIjzORaJ282lTH39GxtFxBa0eaHz
         0v8Q==
X-Gm-Message-State: AOAM533mnUVfQgeOaVtWl6uRpJBcT7L59SPjTXn2uW3nP9LBF2jeHCri
        J8jfGQIPD5xiLJaYtooxGadyKWHRqoQ=
X-Google-Smtp-Source: 
 ABdhPJzB4JhjqKUiQijF4pjiGVjvdSRJOxy7RYQ6tKSTMs1ehMQtVyroivlCYro1QV44ABOEAObA1w==
X-Received: by 2002:a05:6a00:1509:b029:2de:6765:276b with SMTP id
 q9-20020a056a001509b02902de6765276bmr2508738pfu.67.1621339295407;
        Tue, 18 May 2021 05:01:35 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.53])
        by smtp.googlemail.com with ESMTPSA id
 l20sm12757394pjq.38.2021.05.18.05.01.32
        (version=TLS1_2 cipher=ECDHE-ECDSA-AES128-GCM-SHA256 bits=128/128);
        Tue, 18 May 2021 05:01:34 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
        Sean Christopherson <seanjc@google.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>
Subject: [PATCH v4 2/5] KVM: X86: Bail out of direct yield in case of
 under-committed scenarios
Date: Tue, 18 May 2021 05:00:32 -0700
Message-Id: <1621339235-11131-2-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1621339235-11131-1-git-send-email-wanpengli@tencent.com>
References: <1621339235-11131-1-git-send-email-wanpengli@tencent.com>
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Wanpeng Li <wanpengli@tencent.com>

In case of under-committed scenarios, vCPU can get scheduling easily,
kvm_vcpu_yield_to add extra overhead, we can observe a lot of race
between vcpu->ready is true and yield fails due to p->state is
TASK_RUNNING. Let's bail out in such scenarios by checking the length
of current cpu runqueue, it can be treated as a hint of under-committed
instead of guarantee of accuracy. The directed_yield_successful/attempted
ratio can be improved from 50+% to 80+% in the under-committed scenario.

Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
v2 -> v3:
 * update patch description
v1 -> v2:
 * move the check after attempted counting
 * update patch description

 arch/x86/kvm/x86.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 9b6bca616929..dfb7c320581f 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -8360,6 +8360,9 @@ static void kvm_sched_yield(struct kvm_vcpu *vcpu, unsigned long dest_id)
 
 	vcpu->stat.directed_yield_attempted++;
 
+	if (single_task_running())
+		goto no_yield;
+
 	rcu_read_lock();
 	map = rcu_dereference(vcpu->kvm->arch.apic_map);
 

From patchwork Tue May 18 12:00:33 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 12264691
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FORGED_FROMDOMAIN,FREEMAIL_FROM,
	HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT
	autolearn=unavailable autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id D16A5C433B4
	for <kvm@archiver.kernel.org>; Tue, 18 May 2021 12:01:55 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id AD4C36109E
	for <kvm@archiver.kernel.org>; Tue, 18 May 2021 12:01:55 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1349016AbhERMDI (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 18 May 2021 08:03:08 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:49420 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S234375AbhERMC5 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 18 May 2021 08:02:57 -0400
Received: from mail-pf1-x429.google.com (mail-pf1-x429.google.com
 [IPv6:2607:f8b0:4864:20::429])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id EE98AC061573;
        Tue, 18 May 2021 05:01:38 -0700 (PDT)
Received: by mail-pf1-x429.google.com with SMTP id 22so6833731pfv.11;
        Tue, 18 May 2021 05:01:38 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=/2GPjTLQ/YR+wnuKsns5/Hi1QLal2y1jDQCRlyJhbRk=;
        b=ZeIi/gvUNym2Dd9YiXCf0AJ6fBKidFcXrnjVOMnupLUuz68CdlrtfW7q5h4nJIROaL
         TWXnzoNvpXjL+zgx3xqklTashGR+nM8dtmqHEYd6+4EVF13Cy6ZpX6/tws9rkD7dAlOA
         2O4WmVYoYHPa/5omhQZUAT6lXRCI1IZ6H+QVRm208oFVL0GrXO0ZH5ppfkEV8row5IU4
         2NgnNsAgTs/ITFePFDUNowhYlRCJTeUdjAuadljv0cqV/Kgo3tPXNIqfZOJqhnXgsI89
         aa3jreh6+3yBGZMfNp/PE+zNFSlgy26UHE4lSTuxldVNNuORYEyqfksVeXY4pnHyX0Yc
         lVzQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references;
        bh=/2GPjTLQ/YR+wnuKsns5/Hi1QLal2y1jDQCRlyJhbRk=;
        b=ediIZu6qNP2n5peNx8wh2jQi3E+pEC7nOGfWCwWUJ0N34LT03dsYq2PQU6kE5ZxeDo
         Fq+kGmdc4vKiHNU+XAMFYfqBuLGutGRuKuFp0GdXEXk4oGUNIWsf45RmWr1uv88L2tiU
         1kDhUkUjo7uU9jOeTpHUMixj55X1tr/dzWYIH9U3TTVr/GwWXiFLUOgQAUnd3gsJ3AcI
         gq99WHtKQvZx/3s19422/lGeXf7lGZFe67fUdBnfcnzNYDKYgdrF9W0m819cEW3ZF7IW
         LzuvX7RQiNcKFs32B6Aq/3kmsA9TL0KUMRj7lDd84i3EVw5U9Zs6k5StRi1pZsqdqWZx
         FfdQ==
X-Gm-Message-State: AOAM531dCO72/icA8o2kYQVRGcszsBHpq8OeoyIDEYI1np6Wpt/dbFso
        LUWyIIjMXeM/ENBvHKqK4pele5X0IDY=
X-Google-Smtp-Source: 
 ABdhPJxcbPclLPPTmDzgEncZEZSD0AaWP/y/J2B3jrqQOaA8ZnYm4qZvxlrmL+hkSI63xUnGUyL7nA==
X-Received: by 2002:a63:4d5c:: with SMTP id n28mr4700618pgl.436.1621339298356;
        Tue, 18 May 2021 05:01:38 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.53])
        by smtp.googlemail.com with ESMTPSA id
 l20sm12757394pjq.38.2021.05.18.05.01.35
        (version=TLS1_2 cipher=ECDHE-ECDSA-AES128-GCM-SHA256 bits=128/128);
        Tue, 18 May 2021 05:01:37 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
        Sean Christopherson <seanjc@google.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>, stable@vger.kernel.org
Subject: [PATCH v4 3/5] KVM: X86: Fix vCPU preempted state from guest's point
 of view
Date: Tue, 18 May 2021 05:00:33 -0700
Message-Id: <1621339235-11131-3-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1621339235-11131-1-git-send-email-wanpengli@tencent.com>
References: <1621339235-11131-1-git-send-email-wanpengli@tencent.com>
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Wanpeng Li <wanpengli@tencent.com>

Commit 66570e966dd9 (kvm: x86: only provide PV features if enabled in guest's
CPUID) avoids to access pv tlb shootdown host side logic when this pv feature
is not exposed to guest, however, kvm_steal_time.preempted not only leveraged
by pv tlb shootdown logic but also mitigate the lock holder preemption issue.
From guest's point of view, vCPU is always preempted since we lose the reset
of kvm_steal_time.preempted before vmentry if pv tlb shootdown feature is not
exposed. This patch fixes it by clearing kvm_steal_time.preempted before
vmentry.

Fixes: 66570e966dd9 (kvm: x86: only provide PV features if enabled in guest's CPUID)
Reviewed-by: Sean Christopherson <seanjc@google.com>
Cc: stable@vger.kernel.org
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
v1 -> v2:
 * add curly braces

 arch/x86/kvm/x86.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index dfb7c320581f..bed7b5348c0e 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -3105,6 +3105,8 @@ static void record_steal_time(struct kvm_vcpu *vcpu)
 				       st->preempted & KVM_VCPU_FLUSH_TLB);
 		if (xchg(&st->preempted, 0) & KVM_VCPU_FLUSH_TLB)
 			kvm_vcpu_flush_tlb_guest(vcpu);
+	} else {
+		st->preempted = 0;
 	}
 
 	vcpu->arch.st.preempted = 0;

From patchwork Tue May 18 12:00:34 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 12264693
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FORGED_FROMDOMAIN,FREEMAIL_FROM,
	HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 8D932C433B4
	for <kvm@archiver.kernel.org>; Tue, 18 May 2021 12:02:01 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 6D4326109E
	for <kvm@archiver.kernel.org>; Tue, 18 May 2021 12:02:01 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1349025AbhERMDQ (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 18 May 2021 08:03:16 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:49434 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1348967AbhERMC7 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 18 May 2021 08:02:59 -0400
Received: from mail-pj1-x1032.google.com (mail-pj1-x1032.google.com
 [IPv6:2607:f8b0:4864:20::1032])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id A6831C061573;
        Tue, 18 May 2021 05:01:41 -0700 (PDT)
Received: by mail-pj1-x1032.google.com with SMTP id
 pi6-20020a17090b1e46b029015cec51d7cdso1409791pjb.5;
        Tue, 18 May 2021 05:01:41 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=p6jAyFEHVFCxlZGpGxYbrWuXzrXvA1+9+mG4jRlrIQk=;
        b=tqDT9KxazuovNRmwbS4IDOkVHtDo6mjoMa1uRcY3W2LdtNltNuSGmnD9tTl3RSQRPV
         aQ40q/O/ydu3Y+oKrfdL8Grga0tnbpgOaQ/H7F+2BDsN/WI0LNNlNaQ7OSpCDa76b6Zt
         zNIPVabi3l57HxgDO1vD/JsZWAZ47QxOOfEm9NrNrxsXoLQbYTm8pX1ag0ODn9yviFCx
         iLXc4NAHQqB/9TpyZ12s/FfjZLVLOia3t/kfxgqSrdOMzLmJLR2DThE3+NMd54ITF8us
         0zuq6SW+9Hf0wB9ZaKI4HMvSFxK8Q08syOS6UtQcM7P3MJmdfqM4e/6rkIDWPMNCE13E
         t8Hw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references;
        bh=p6jAyFEHVFCxlZGpGxYbrWuXzrXvA1+9+mG4jRlrIQk=;
        b=RCdIfmXn7beW5k4nJPNKJGKCK0+Dxoga/D/EF60MDE5TTn843KzJtV0X/2yoHMnHPx
         cEGhghxDJziwVAPEjUz8JPQwHOYTK7bKO9wos3sP1pJT9XlAhhayhgykxpgTsdpn5IwT
         bLK+3fQVjNQEQbB5pooswYYPLhhxY6wHygR9RPJ17OqOA8zgWtUikQ+MPQG0I3pNkVZ9
         mTe/WVDpurGZ5UVE6/pzUlvAYadZZntaNrmxPvrXral5/Fo5rD2IBTRU7ZeL+luUTjL6
         hEXgHsvPFSu/ub+szPZtjb4lyuw+Y4ogAzmICl/5emSsS033UqRw+2C6C41gQkZ3WZw5
         b3Lg==
X-Gm-Message-State: AOAM530DLRnn56OvMbSfwsNtPKYPXd3pzsBs0nAd5iXV0ncmzgk69LzD
        fh4MJa7DppRU1XUFoAxGq/7teA1O/sk=
X-Google-Smtp-Source: 
 ABdhPJx3RfT9dV9jIXyfohnH4+VlEo/N3L8V1FP0l3Eoi6Fi1Sa88+ub7ujncc8i0vPmhKB89Z8WYQ==
X-Received: by 2002:a17:90b:ed5:: with SMTP id
 gz21mr5200301pjb.102.1621339301043;
        Tue, 18 May 2021 05:01:41 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.53])
        by smtp.googlemail.com with ESMTPSA id
 l20sm12757394pjq.38.2021.05.18.05.01.38
        (version=TLS1_2 cipher=ECDHE-ECDSA-AES128-GCM-SHA256 bits=128/128);
        Tue, 18 May 2021 05:01:40 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
        Sean Christopherson <seanjc@google.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>
Subject: [PATCH v4 4/5] KVM: X86: hyper-v: Task srcu lock when accessing
 kvm_memslots()
Date: Tue, 18 May 2021 05:00:34 -0700
Message-Id: <1621339235-11131-4-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1621339235-11131-1-git-send-email-wanpengli@tencent.com>
References: <1621339235-11131-1-git-send-email-wanpengli@tencent.com>
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Wanpeng Li <wanpengli@tencent.com>

   WARNING: suspicious RCU usage
   5.13.0-rc1 #4 Not tainted
   -----------------------------
   ./include/linux/kvm_host.h:710 suspicious rcu_dereference_check() usage!

  other info that might help us debug this:

  rcu_scheduler_active = 2, debug_locks = 1
   1 lock held by hyperv_clock/8318:
    #0: ffffb6b8cb05a7d8 (&hv->hv_lock){+.+.}-{3:3}, at: kvm_hv_invalidate_tsc_page+0x3e/0xa0 [kvm]

  stack backtrace:
  CPU: 3 PID: 8318 Comm: hyperv_clock Not tainted 5.13.0-rc1 #4
  Call Trace:
   dump_stack+0x87/0xb7
   lockdep_rcu_suspicious+0xce/0xf0
   kvm_write_guest_page+0x1c1/0x1d0 [kvm]
   kvm_write_guest+0x50/0x90 [kvm]
   kvm_hv_invalidate_tsc_page+0x79/0xa0 [kvm]
   kvm_gen_update_masterclock+0x1d/0x110 [kvm]
   kvm_arch_vm_ioctl+0x2a7/0xc50 [kvm]
   kvm_vm_ioctl+0x123/0x11d0 [kvm]
   __x64_sys_ioctl+0x3ed/0x9d0
   do_syscall_64+0x3d/0x80
   entry_SYSCALL_64_after_hwframe+0x44/0xae

kvm_memslots() will be called by kvm_write_guest(), so we should take the srcu lock.

Fixes: e880c6ea5 (KVM: x86: hyper-v: Prevent using not-yet-updated TSC page by secondary CPUs)
Reviewed-by: Vitaly Kuznetsov <vkuznets@redhat.com>
Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
---
 arch/x86/kvm/hyperv.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/arch/x86/kvm/hyperv.c b/arch/x86/kvm/hyperv.c
index f98370a39936..f00830e5202f 100644
--- a/arch/x86/kvm/hyperv.c
+++ b/arch/x86/kvm/hyperv.c
@@ -1172,6 +1172,7 @@ void kvm_hv_invalidate_tsc_page(struct kvm *kvm)
 {
 	struct kvm_hv *hv = to_kvm_hv(kvm);
 	u64 gfn;
+	int idx;
 
 	if (hv->hv_tsc_page_status == HV_TSC_PAGE_BROKEN ||
 	    hv->hv_tsc_page_status == HV_TSC_PAGE_UNSET ||
@@ -1190,9 +1191,16 @@ void kvm_hv_invalidate_tsc_page(struct kvm *kvm)
 	gfn = hv->hv_tsc_page >> HV_X64_MSR_TSC_REFERENCE_ADDRESS_SHIFT;
 
 	hv->tsc_ref.tsc_sequence = 0;
+
+	/*
+	 * Take the srcu lock as memslots will be accessed to check the gfn
+	 * cache generation against the memslots generation.
+	 */
+	idx = srcu_read_lock(&kvm->srcu);
 	if (kvm_write_guest(kvm, gfn_to_gpa(gfn),
 			    &hv->tsc_ref, sizeof(hv->tsc_ref.tsc_sequence)))
 		hv->hv_tsc_page_status = HV_TSC_PAGE_BROKEN;
+	srcu_read_unlock(&kvm->srcu, idx);
 
 out_unlock:
 	mutex_unlock(&hv->hv_lock);

From patchwork Tue May 18 12:00:35 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Wanpeng Li <kernellwp@gmail.com>
X-Patchwork-Id: 12264695
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.8 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FORGED_FROMDOMAIN,FREEMAIL_FROM,
	HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,INCLUDES_PATCH,
	MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,USER_AGENT_GIT autolearn=ham
	autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id D8200C433B4
	for <kvm@archiver.kernel.org>; Tue, 18 May 2021 12:02:04 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id BE8A46124C
	for <kvm@archiver.kernel.org>; Tue, 18 May 2021 12:02:04 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1349003AbhERMDU (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 18 May 2021 08:03:20 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:49454 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1348997AbhERMDC (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 18 May 2021 08:03:02 -0400
Received: from mail-pf1-x435.google.com (mail-pf1-x435.google.com
 [IPv6:2607:f8b0:4864:20::435])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 5E400C061761;
        Tue, 18 May 2021 05:01:44 -0700 (PDT)
Received: by mail-pf1-x435.google.com with SMTP id b13so3477992pfv.4;
        Tue, 18 May 2021 05:01:44 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=from:to:cc:subject:date:message-id:in-reply-to:references;
        bh=soOUzbLJHd79dLzXUneFH08UM8yTlNutja4acLt0+qU=;
        b=hIdVLuZbbk0swz4hpVBO5hP1ydSkuNP4SwlXYVJ4f2xrIqZy7KGaH9grl6v6LlNJLY
         bDirjnkwMnbG96k6JaFACCjjPV3taK7feKB/DCqVY8eq4jpnFPXi3V5pOs3ZevSSzdaj
         qXSii3s1GknciQzSYzAYHIsP951Uzqo6egApt9F8IVg0MfgNjio+FwycWiJSDO7nCHOv
         08n40f/UU7WaAZY5IPClwkxtiRyiFD4z2Mh3id4MXy4nuKcyMZIqm0jl1uMxa4uYv0p1
         ljY1NyOs1CbzGigFkjqB2GHMVfqaiPTF8oB7984LCCzOTS/8795Ug1xMRoddTNoyU2BW
         K+Kg==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:from:to:cc:subject:date:message-id:in-reply-to
         :references;
        bh=soOUzbLJHd79dLzXUneFH08UM8yTlNutja4acLt0+qU=;
        b=ncurSObIRNDrVewa6ONUlxk1LzWx1mCvkmL6Opcn9ytT/LM3eUi7WJCW3/lnBlnHz5
         uMf5BRmw/a+FjG45hFa4gb06RoabrO1yM/FLGHEbafSyV+YkUN0LVEVxS4J3nW4xxdPc
         Opq3lCK+Lpsf9imhArM6mSEUAlzr/LgWhMu/7h3XJmQ1Fze+bGnqkaCHaY5diD2usIHA
         l7ojQ120r+Y5jUMN029Rn4PmPAsIrktN/SA5lOKQ2NLBuC0lMM+WABKFL8oUNm/P3E1I
         z0KYAZPjhvcUPghWhidaVhJQobSZvN/GXbpZWKvTOrKi0H1GBV7pEzzr5WB8uVDRWBFH
         YQMQ==
X-Gm-Message-State: AOAM533lD7Rec7V1kmH4RXBlvIPz5AYZWGkMufDTgfiXRdusPkLUOxIW
        uNlHJwburaF8y3sY3oBLabvkjfGc0tw=
X-Google-Smtp-Source: 
 ABdhPJztev0nvWB7p2r+uTPVQkH9uSr4620Rogko2ex6afamBnUY4Y+8h39mp42uX4FtDw+3fhQjsQ==
X-Received: by 2002:a63:d744:: with SMTP id w4mr4791316pgi.76.1621339303764;
        Tue, 18 May 2021 05:01:43 -0700 (PDT)
Received: from localhost.localdomain ([203.205.141.53])
        by smtp.googlemail.com with ESMTPSA id
 l20sm12757394pjq.38.2021.05.18.05.01.41
        (version=TLS1_2 cipher=ECDHE-ECDSA-AES128-GCM-SHA256 bits=128/128);
        Tue, 18 May 2021 05:01:43 -0700 (PDT)
From: Wanpeng Li <kernellwp@gmail.com>
X-Google-Original-From: Wanpeng Li <wanpengli@tencent.com>
To: linux-kernel@vger.kernel.org, kvm@vger.kernel.org
Cc: Paolo Bonzini <pbonzini@redhat.com>,
        Sean Christopherson <seanjc@google.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>
Subject: [PATCH v4 5/5] KVM: LAPIC: Narrow the timer latency between
 wait_lapic_expire and world switch
Date: Tue, 18 May 2021 05:00:35 -0700
Message-Id: <1621339235-11131-5-git-send-email-wanpengli@tencent.com>
X-Mailer: git-send-email 2.7.4
In-Reply-To: <1621339235-11131-1-git-send-email-wanpengli@tencent.com>
References: <1621339235-11131-1-git-send-email-wanpengli@tencent.com>
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

From: Wanpeng Li <wanpengli@tencent.com>

Let's treat lapic_timer_advance_ns automatic tuning logic as hypervisor
overhead, move it before wait_lapic_expire instead of between wait_lapic_expire 
and the world switch, the wait duration should be calculated by the 
up-to-date guest_tsc after the overhead of automatic tuning logic. This 
patch reduces ~30+ cycles for kvm-unit-tests/tscdeadline-latency when testing 
busy waits.

Signed-off-by: Wanpeng Li <wanpengli@tencent.com>
Reviewed-by: Sean Christopherson <seanjc@google.com>
---
v3 -> v4:
 * consider automatic tuning is disabled and timer did not arrive early
 * add a code comment

 arch/x86/kvm/lapic.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/arch/x86/kvm/lapic.c b/arch/x86/kvm/lapic.c
index c0ebef560bd1..5d91f2367c31 100644
--- a/arch/x86/kvm/lapic.c
+++ b/arch/x86/kvm/lapic.c
@@ -1598,11 +1598,19 @@ static void __kvm_wait_lapic_expire(struct kvm_vcpu *vcpu)
 	guest_tsc = kvm_read_l1_tsc(vcpu, rdtsc());
 	apic->lapic_timer.advance_expire_delta = guest_tsc - tsc_deadline;
 
+	if (lapic_timer_advance_dynamic) {
+		adjust_lapic_timer_advance(vcpu, apic->lapic_timer.advance_expire_delta);
+		/*
+		 * If the timer fired early, reread the TSC to account for the
+		 * overhead of the above adjustment to avoid waiting longer
+		 * than is necessary.
+		 */
+		if (guest_tsc < tsc_deadline)
+			guest_tsc = kvm_read_l1_tsc(vcpu, rdtsc());
+	}
+
 	if (guest_tsc < tsc_deadline)
 		__wait_lapic_expire(vcpu, tsc_deadline - guest_tsc);
-
-	if (lapic_timer_advance_dynamic)
-		adjust_lapic_timer_advance(vcpu, apic->lapic_timer.advance_expire_delta);
 }
 
 void kvm_wait_lapic_expire(struct kvm_vcpu *vcpu)
